README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# podcleaner

<!-- badges: start -->

<!-- badges: end -->

The Scottish Post Office directories are annual directories that provide
an alphabetical list of a town’s or county’s inhabitants including their
forename, surname, occupation and address(es); they provide a solid
basis for researching Scotland’s family, trade, and town history. A
large number of these, covering most of Scotland and dating from 1773 to
1911, can be accessed in digitised form from the [National Library of
Scotland](https://digital.nls.uk/directories/). `podcleaner` attempts to
clean optical character recognition (OCR) errors in directory records
after they’ve been parsed and saved to “csv” files using a third party
tool\[1\]. The package further attempts to match records from trades and
general directories. See the tests folder for examples running
unexported functions.

## Load

Load general and trades directory samples in memory from “csv” files:

  - **Globals**

<!-- end list -->

``` r
library(podcleaner)

directories <- c("1861-1862")

progress <- TRUE; verbose <- FALSE
```

  - **General directories**

<!-- end list -->

``` r
path_directories <- utils_make_path("data", "general-directories")

general_directory <- utils_load_directories_csv(
  type = "general", directories, path_directories, verbose
)

print.data.frame(general_directory)
#>   directory page    surname forename
#> 1 1861-1862   71       ABOT      Wm.
#> 2 1861-1862   71 ABRCROMBIE     Alex
#>                                                occupation
#> 1 Wine and spirit mercht — See Advertisement in Appendix.
#> 2                                                        
#>                                                    addresses
#> 1                           1S20 Londn rd; ho. 13<J Queun sq
#> 2 Bkr; I2 Dixon Street, & 29 Auderstn Qu.; res 2G5 Argul st.
```

  - **Trades directories**

<!-- end list -->

``` r
path_directories <- utils_make_path("data", "trades-directories")

trades_directory <- utils_load_directories_csv(
  type = "trades", directories, path_directories, verbose
)

print.data.frame(trades_directory)
#>   directory page rank                                              occupation
#> 1 1861-1862   71  135 Wine and spirit mercht — See Advertisement in Appendix.
#> 2 1861-1862   71  326                                                     Bkr
#> 3 1861-1862   71  586                                               Victualer
#>          type    surname forename address.trade.body address.trade.number
#> 1 OWN ACCOUNT       ABOT      Wm.          Londn rd.                 1S20
#> 2 OWN ACCOUNT ABRCROMBIE     Alex           Dixen pl                   I2
#> 3 OWN ACCOUNT       BLAI  Jon Hug           High St.                  2S0
```

## Clean

Clean records from both datasets:

  - **General directories**

<!-- end list -->

``` r
general_directory <-
  general_clean_directory(general_directory, progress, verbose)

print.data.frame(general_directory)
#>   directory page    surname  forename               occupation
#> 1 1861-1862   71     Abbott   William Wine and spirit merchant
#> 2 1861-1862   71 Abercromby Alexander                    Baker
#> 3 1861-1862   71 Abercromby Alexander                    Baker
#>   address.trade.number address.trade.body address.house.number
#> 1               18, 20       London Road.                  136
#> 2                   12      Dixon Street.                  265
#> 3                   29    Anderston Quay.                  265
#>   address.house.body
#> 1      Queen Square.
#> 2     Argyle Street.
#> 3     Argyle Street.
```

  - **Trades directories**

<!-- end list -->

``` r
trades_directory <-
  trades_clean_directory(trades_directory, progress, verbose)

print.data.frame(trades_directory)
#>   directory page rank    surname  forename               occupation        type
#> 1 1861-1862   71  135     Abbott   William Wine and spirit merchant OWN ACCOUNT
#> 2 1861-1862   71  326 Abercromby Alexander                    Baker OWN ACCOUNT
#> 3 1861-1862   71  586      Blair John Hugh               Victualler OWN ACCOUNT
#>   address.trade.number address.trade.body
#> 1               18, 20       London Road.
#> 2                   12       Dixon Place.
#> 3                  280       High Street.
```

## Match

Match general to trades directory records:

``` r
distance <- TRUE; matches <- TRUE

directory <- combine_match_general_to_trades(
  trades_directory, general_directory, progress, verbose, distance, matches,
  method = "osa", max_dist = 5L
)

print.data.frame(directory)
#>   directory page rank    surname  forename               occupation        type
#> 1 1861-1862   71  135     Abbott   William Wine and spirit merchant OWN ACCOUNT
#> 2 1861-1862   71  326 Abercromby Alexander                    Baker OWN ACCOUNT
#> 3 1861-1862   71  586      Blair John Hugh               Victualler OWN ACCOUNT
#>   address.trade.number address.trade.body address.house.number
#> 1               18, 20       London Road.                  136
#> 2                   12       Dixon Place.                  265
#> 3                  280       High Street.                     
#>                       address.house.body distance
#> 1                          Queen Square.        0
#> 2                         Argyle Street.        5
#> 3 Failed to match with general directory       NA
#>                                     match
#> 1    Abbott William - 18, 20, London Road
#> 2 Abercromby Alexander - 12, Dixon Street
#> 3                                    <NA>
```

Directory records are compared and eventually matched using a distance
metric calculated with the method and corresponding parameters specified
in arguments. Under the hood the
[fuzzyjoin](https://www.rdocumentation.org/packages/fuzzyjoin/versions/0.1.6)
package and the
[stringdist\_left\_join](https://www.rdocumentation.org/packages/fuzzyjoin/versions/0.1.6/topics/stringdist_join)
function in particular, help with the matching operations.

## Save

``` r
utils_IO_write(directory, "dev", "post-office-directory")
```

1.  See for example the python
    [podparser](https://pythonhosted.org/podparser/) library.


NEWS.md
# podcleaner 0.1.2

## New parameters for combine_match_general_to_trades

* `distance`: boolean flag specifying whether (`TRUE`) or not (`FALSE`) a column 
'distance' showing the string distance between records used for their matching 
and calculated using the method specified below should be added to the output 
dataset.  
* `matches`: boolean flag specifying whether (`TRUE`) or not (`FALSE`) a column 
'match' showing general directory matches' name and address(es) should be added 
to the output dataset.


# podcleaner 0.1.1

## Bug fixes

* globals_regex_house_to_address becomes ";\\s\\b(?:[bh][op](?:use)?|res(?:idence)?)\\b\\.?.*".  
* globals_regex_house_split_trade becomes: "(?:^|[;,\u201e\\s]*)(?<![\\-])?\\b(?:res(?:id)?(?:ence)?|(?:(?:[bdht]|li|jh)[aop])(?:[ui\\/]se)?s?)\\b[.,\u201e\\s]+".


# podcleaner 0.1.0

* 2021-20-28

First submission to CRAN.


tests/testthat/test_clean.r
# address_body ####
test_that("clean_address_body works for various addresses", {
  expect_equal(
    clean_address_body(c("18, 20 Londun st.", "136 Dixon rd")),
    c("18, 20 London Street.", "136 Dixon Road.")
  )
})

# address_number ####
test_that("clean_address_number works for various addresses", {
  expect_equal(
    clean_address_number(c("1S20 London Street.", "13<J Dixon Road.")),
    c("18, 20 London Street", "136 Dixon Road")
  )
})

# title ####
test_that("clean_title works for various titles", {
  expect_equal(
    clean_title(c("  Capt William Abbott", "?-Rev Alexander Abercromby")),
    c("Captain William Abbott", "Reverend Alexander Abercromby")
  )
})

# forename ####
test_that("clean_forename works for various forenames", {
  expect_equal(clean_forename(c("McWm.", "Jas.")), c("Mac William", "James"))
})

# surname ####
test_that("clean_surname works for various surnames", {
  expect_equal(clean_surname(c("ABOT", "ABRCROMBIE")), c("Abbott", "Abercromby"))
})

# specials ####
test_that("clean_specials works for various special characters", {
  expect_equal(
    clean_specials(c("Wine and spirit \u00bbmerchant", "Mac\u00bb William")),
    c("Wine and spirit merchant", "Mac William")
  )
})

# parentheses ####
test_that("clean_parentheses works in general", {
  expect_equal(clean_parentheses("Miller Street. (Agents)."), c("Miller Street."))
})

# mac ####
test_that("clean_mac works in general", {
  expect_equal(clean_mac("McWilliam"), c("Mac William"))
})

# address_pre_clean ####
test_that("clean_address_pre_clean works for various addresses", {
  expect_equal(
    clean_address_pre_clean(c(": 1S20 Lond»n st.   -", "13<J st enoch sq,;")),
    c("1S20 Londn st.", "13<J Saint Enoch sq.")
  )
})

# address_post_clean ####
test_that("clean_address_post_clean works for various addresses", {
  expect_equal(
    clean_address_post_clean(c(": 18, 20 London Street.   -", "136 Saint Enoch Square,;")),
    c("18, 20 London Street.", "136 Saint Enoch Square.")
  )
})

# address_mac ####
test_that("clean_address_mac works for various addresses", {
  expect_equal(
    clean_address_mac(c("McWilliam", "M'Donald", "M c Fyfe")),
    c("Mac William", "Mac Donald", "Mac Fyfe")
  )
})

# address_saints ####
test_that("clean_address_saints works for various addresses", {
  expect_equal(
    clean_address_saints(c("St Georges st.", "St Enoch sq")),
    c("Saint George st.", "Saint Enoch sq")
  )
})

# address_possessives ####
test_that("clean_address_possessives works for various address bodies", {
  expect_equal(
    clean_address_possessives(c("Adam's Court", "Aird's Lane")),
    c("Adams Court", "Airds Lane")
  )
})

# address_ends ####
test_that("clean_address_ends works for various addresses", {
  expect_equal(
    clean_address_ends(c(" -; 18, 20 London st.", ",,12 Dixon st.$  ")),
    c("18, 20 London st.", "12 Dixon st.")
  )
})

# address_attached_words ####
test_that("clean_address_attached_words works for various addresses", {
  expect_equal(
    clean_address_attached_words(c("18, 20 LondonSt.", "12 Dixon.St.")),
    c("18, 20 London St.", "12 Dixon St.")
  )
})

# address_places ####
test_that("clean_address_places works for various address bodies", {
  expect_equal(
    clean_address_places(c("Bothwell Cir.", "Railway arch.")),
    c("Bothwell Circus.", "Railway Arches.")
  )
})

# address_worksites ####
test_that("clean_address_worksites works in general", {
  expect_equal(clean_address_worksites("Woodyd"), "Woodyard")
})

# address_suffixes ####
test_that("clean_address_suffixes works in general", {
  expect_equal(
    clean_address_suffixes(c("36 Rose street so")), "36 Rose street South"
  )
})

# address_names ####
test_that("clean_address_names works for various addresses", {
  expect_equal(
    clean_address_names(c("18, 20 Londun Street", "136 Dixn Road")),
    c("18, 20 London Street", "136 Dixon Road")
  )
})

# address_others ####
test_that("clean_address_others works for various addresses", {
  expect_equal(
    clean_address_others(c("18, 20 London. Street, Agents", "136 Dixon Road d")),
    c("18, 20 London Street", "136 Dixon Road")
  )
})

# name_ends ####
test_that("clean_name_ends works for various names", {
  expect_equal(
    clean_name_ends(c("William-", "$Alexander", "John Hugh   ")),
    c("William", "Alexander", "John Hugh")
  )
})

# forename_separate_words ####
test_that("clean_forename_separate_words works for various forenames", {
  expect_equal(
    clean_forename_separate_words(c("William", "Alexander", "JohnHugh")),
    c("William", "Alexander", "John Hugh")
  )
})

# forename_punctuation ####
test_that("clean_forename_punctuation works for various forenames", {
  expect_equal(
    clean_forename_punctuation(c("William;", "Miss", "John,Hugh")),
    c("William.", "Miss.", "John Hugh.")
  )
})

# forename_spelling ####
test_that("clean_forename_spelling works for various forenames", {
  expect_equal(
    clean_forename_spelling(c("Wm", "A.", "Jn Huh")),
    c("William", "Alexander", "John Hugh")
  )
})

# surname_punctuation ####
test_that("clean_surname_punctuation works for various surnames", {
  expect_equal(
    clean_surname_punctuation(c("Abbott*", "*Abercromby", "Blair")),
    c("Abbott", "Abercromby", "Blair")
  )
})

# surname_spelling ####
test_that("clean_surname_spelling works for various surnames", {
  expect_equal(
    clean_surname_spelling(c("ABOT junior", "ABRCROMBIE", "BLAI senior")),
    c("Abbott", "Abercromby", "Blair")
  )
})

# string_ends ####
test_that("clean_string_ends works for various strings", {
  expect_equal(
    clean_string_ends(c(" -; 18, 20 Mary hill.*", ",,12 Dixon Street^")),
    c("18, 20 Mary hill", "12 Dixon Street")
  )
})



tests/testthat/test_trades.r
# clean_entries ####
test_that("trades_clean_entries works for various entries", {
  directory <- data.frame(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("ABOT", "ABRCROMBIE", "BLAI"),
    forename = c("Wm.", "Alex", "Jon Hug"),
    occupation = c(
      "Wine and spirit mercht — See Advertisement in Appendix.", "Bkr",
      "Victualer"
    ),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("1S20", "I2", "2S0"),
    address.trade.body = c("Londn rd.", "Dixen pl", "High St."),
    stringsAsFactors = FALSE
  )
  out <- data.frame(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("Abbott", "Abercromby", "Blair"),
    forename = c("William", "Alexander", "John Hugh"),
    occupation = c("Wine and spirit merchant", "Baker", "Victualler"),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("18, 20", "12", "280"),
    address.trade.body = c("London Road.", "Dixon Place.", "High Street."),
    stringsAsFactors = FALSE
  )
  expect_equal(trades_clean_entries(directory, verbose = FALSE), out)
})


# clean_directory ####
test_that("trades_clean_directory works in general", {
  directory <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("ABOT", "ABRCROMBIE", "BLAI"),
    forename = c("Wm.", "Alex", "Jon Hug"),
    occupation = c(
      "Wine and spirit mercht — See Advertisement in Appendix.", "Bkr",
      "Victualer"
    ),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("1S20", "I2", "2S0"),
    address.trade.body = c("Londn rd.", "Dixen pl", "High St.")
  )
  out <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("Abbott", "Abercromby", "Blair"),
    forename = c("William", "Alexander", "John Hugh"),
    occupation = c("Wine and spirit merchant", "Baker", "Victualler"),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("18, 20", "12", "280"),
    address.trade.body = c("London Road.", "Dixon Place.", "High Street."),
  )
  expect_equal(trades_clean_directory(directory, verbose = FALSE), out)
})

# trades_clean_directory_plain ####
test_that("trades_clean_directory_plain works in general", {
  directory <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("ABOT", "ABRCROMBIE", "BLAI"),
    forename = c("Wm.", "Alex", "Jon Hug"),
    occupation = c(
      "Wine and spirit mercht — See Advertisement in Appendix.", "Bkr",
      "Victualer"
    ),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("1S20", "I2", "2S0"),
    address.trade.body = c("Londn rd.", "Dixen pl", "High St.")
  )
  out <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("Abbott", "Abercromby", "Blair"),
    forename = c("William", "Alexander", "John Hugh"),
    occupation = c("Wine and spirit merchant", "Baker", "Victualler"),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("18, 20", "12", "280"),
    address.trade.body = c("London Road.", "Dixon Place.", "High Street."),
  )
  expect_equal(trades_clean_directory_plain(directory, verbose = FALSE), out)
})

# trades_clean_directory_progress ####
test_that("trades_clean_directory_progress works in general", {
  directory <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("ABOT", "ABRCROMBIE", "BLAI"),
    forename = c("Wm.", "Alex", "Jon Hug"),
    occupation = c(
      "Wine and spirit mercht — See Advertisement in Appendix.", "Bkr",
      "Victualer"
    ),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("1S20", "I2", "2S0"),
    address.trade.body = c("Londn rd.", "Dixen pl", "High St.")
  )
  out <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("Abbott", "Abercromby", "Blair"),
    forename = c("William", "Alexander", "John Hugh"),
    occupation = c("Wine and spirit merchant", "Baker", "Victualler"),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("18, 20", "12", "280"),
    address.trade.body = c("London Road.", "Dixon Place.", "High Street."),
  )
  expect_equal(trades_clean_directory_progress(directory, verbose = FALSE), out)
})

# clean_directory ####
test_that("trades_clean_directory works in general", {
  directory <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("ABOT", "ABRCROMBIE", "BLAI"),
    forename = c("Wm.", "Alex", "Jon Hug"),
    occupation = c(
      "Wine and spirit mercht — See Advertisement in Appendix.", "Bkr",
      "Victualer"
    ),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("1S20", "I2", "2S0"),
    address.trade.body = c("Londn rd.", "Dixen pl", "High St.")
  )
  out <- tibble::tibble(
    page = rep("71", 3L),
    rank = c("135", "326", "586"),
    surname = c("Abbott", "Abercromby", "Blair"),
    forename = c("William", "Alexander", "John Hugh"),
    occupation = c("Wine and spirit merchant", "Baker", "Victualler"),
    type = rep("OWN ACCOUNT", 3L),
    address.trade.number = c("18, 20", "12", "280"),
    address.trade.body = c("London Road.", "Dixon Place.", "High Street."),
  )
  expect_equal(
    trades_clean_directory(directory, progress = TRUE, verbose = FALSE), out
  )
})


R/combine.r
# Define globals ####
`%>%` <- magrittr::`%>%`
ignore_case <- TRUE
perl <- TRUE


# combine_random_string_if_pattern ####

#' Conditionally return a random string
#'
#' Search for specified pattern in provided string; if found returns a 22
#'   character long random string otherwise return original string.
#'
#' @param string A character string.
#' @param regex Character string regex specifying the pattern to look for in
#'   `string`.
#'
#' @return A length 1 character string vector: 22 character long random string
#'   if `regex` found in `string`, `string` otherwise.
combine_random_string_if_pattern <- function(string, regex){
  ifelse(
    grepl(regex, string, ignore_case, perl),
    stringi::stri_rand_strings(1L, 22L), string
  )
}


# combine_random_string_if_no_address ####

#' Conditionally return a random string
#'
#' Returns a 22 character long random string if address provided is labelled as
#'   missing ("No trade/house address found").
#'
#' @param address A character string.
#'
#' @return A length 1 character string vector: 22 character long random string
#'   if `address` labelled as missing ("No trade/house address found"),
#'   `address` otherwise.
combine_random_string_if_no_address <- function(address){
  regex <- "^No.+address\\sfound$"
  combine_random_string_if_pattern(address, regex)
}


# combine_no_trade_address_to_random_string ####

#' Mutate operation(s) in directory data.frame address.trade column.
#'
#' Replaces missing trade address(es) in the provided Scottish post office
#'   directory data.frame with random string(s). Random string(s) only show(s)
#'   in body of trade address entry(/ies).
#'
#' @param directory A Scottish post office directory in the form of a data.frame
#'   or other object that inherits from the data.frame class such as a
#'   \code{\link[tibble]{tibble}}. Columns must at least include `address.trade`.
#'
#' @return A data.frame of the same class as the one provided in `directory`;
#'   columns include at least `address.trade`.
#'
#' @section Details:
#' Prevents unwarranted matches when matching general to trades directory.
#'   Unrelated records with similar name and trade address entry labelled as
#'   missing would be otherwise matched.
combine_no_trade_address_to_random_string <- function(directory){
  address.trade <- NULL

  dplyr::mutate(
    directory,
    address.trade = purrr::map_chr(
      address.trade, ~ combine_random_string_if_no_address(.x)
    )
  )
}


# combine_make_match_string ####

#' Mutate operation(s) in directory data.frame trade address column
#'
#' Creates a 'match.string' column in the provided Scottish post office
#'   directory data.frame composed of entry(/ies) full name and trade address
#'   pasted together. Missing trade address entry(/ies) are replaced with a
#'   random generated string.
#'
#' @param directory A Scottish post office directory in the form of a data.frame
#'   or other object that inherits from the data.frame class such as a
#'   \code{\link[tibble]{tibble}}. Columns must at least include `forename`,
#'   `surname`, `address.trade.number`, `address.trade.body`.
#'
#' @return A data.frame of the same class as the one provided in `directory`;
#'   columns include at least `forename`, `surname`, `address.trade.number`,
#'   `address.trade.body`, `match.string`.
#'
#' @section Details:
#' The purpose of the 'match.string' column is to facilitates the matching of
#'   the general to trades directory down the line. It allows to calculate a
#'   string distance metric between each pair of entries and match those falling
#'   below a specified threshold.
#'
#' @seealso \code{\link{combine_match_general_to_trades}} for the matching of
#'   the general to trades directory.
combine_make_match_string <- function(directory){
  address.trade <- name <- NULL

  tidyr::unite(
    directory,
    "address.trade", dplyr::matches("address.trade"),
    sep = ", ", remove = FALSE, na.rm = TRUE
  ) %>%
    tidyr::unite(
      "name", dplyr::matches("name$"), sep = " ", remove = FALSE, na.rm = TRUE
    ) %>%
    utils_clean_ends(name, address.trade) %>%
    combine_no_trade_address_to_random_string() %>%
    tidyr::unite(
      "match.string", c(dplyr::matches("^name"), "address.trade"), sep = " - ",
      remove = FALSE, na.rm = TRUE
    ) %>%
    dplyr::select(-c(address.trade, name))
}


# combine_has_match_failed ####

#' Check for failed matches
#'
#' Provided with two equal length vectors, returns TRUE for indexes where both
#'   entries are "NA" and FALSE otherwise.
#'
#' @param number A vector of address number(s). Integer or character string.
#' @param body A character string vector of address body(/ies).
#'
#' @return A boolean vector: `TRUE` for indexes where both
#'   `number` and `body` are "NA", `FALSE` otherwise.
combine_has_match_failed <- function(number, body){
  (is.na(number) & is.na(body))
}


# combine_label_if_match_failed ####

#' Label failed matches
#'
#' Labels failed matches as such.
#'
#' @param type A Character string, one of: "number" or "body". Type of column to
#'   label.
#' @param ... Further arguments to be passed down to
#'   \code{\link{combine_has_match_failed}}
#'
#' @return A character string vector: address(es) "number" or "body" as
#'   specified in `type` if match succeeded, "" (type = "number") or
#'   "Failed to match with general directory" (type = "body") otherwise.
combine_label_if_match_failed <- function(type = c("number", "body"), ...){
  txt <- switch(
    type, "number" = "", "body" = "Failed to match with general directory"
  )
  args <- list(...)
  dplyr::if_else(combine_has_match_failed(...), txt, args[[type]])
}


# combine_get_address_house_type ####

#' Get house address column type
#'
#' Identifies the type of the house address column provided: number or body.
#'
#' @param column A Character string: ends in "house.number" or "house.body".
#'
#' @return A Character string: "number" or "body".
combine_get_address_house_type <- function(column){
  regmatches(
    column, regexpr(globals_regex_get_address_house_type, column, perl = perl)
  )
}


# combine_label_failed_matches ####

#' Label failed matches
#'
#' Labels failed matches as such in the provided Scottish post office directory
#'   data.frame.
#'
#' @param directory A Scottish post office directory in the form of a data.frame
#'   or other object that inherits from the data.frame class such as a
#'   \code{\link[tibble]{tibble}}. Columns must at least include
#'   `address.house.number`, `address.house.body`.
#'
#' @return A data.frame of the same class as the one provided in `directory`.
#'   Columns include `address.house.number`, `address.house.body`. For entries
#'   for which both `address.house.number` and `address.house.body` are `NA`,
#'   `address.house.number` and `address.house.body` are labelled as "" and
#'   "Failed to match with general directory" respectively.
combine_label_failed_matches <- function(directory){
  dplyr::mutate(
    directory,
    dplyr::across(
      .cols = dplyr::matches(globals_regex_address_house_body_number),
      ~ combine_label_if_match_failed(
        type = combine_get_address_house_type(dplyr::cur_column()),
        number = address.house.number, body = address.house.body
      )
    )
  )
}


# combine_match_general_to_trades_plain ####

#' Match general to trades directory records
#'
#' Attempts to complement Scottish post office trades directory data.frame with
#'   house address information from the Scottish post office general directory
#'   data.frame provided by matching records from the two datasets using the
#'   distance metric specified.
#'
#' @param trades_directory A Scottish post office trades directory in the form
#'   of a data.frame or other object that inherits from the data.frame class
#'   such as a \code{\link[tibble]{tibble}}. Columns must at least include
#'   `surname`, `forename`, `address.trade.number`, `address.trade.body`.
#' @param general_directory A Scottish post office general directory in the form
#'   of a data.frame or other object that inherits from the data.frame class
#'   such as a \code{\link[tibble]{tibble}}. Columns must at least include
#'   `surname`, `forename`, `address.trade.number`, `address.trade.body`,
#'   `address.house.number`, `address.house.body`.
#' @param matches Whether (`TRUE`) or not (`FALSE`) a column 'match' showing
#'   general directory matches' name and address(es) should be added to the
#'   output dataset.
#' @param verbose Whether the function should be executed silently (`FALSE`) or
#'   not (`TRUE`).
#' @param ... Further arguments to be passed down to
#'   \code{\link[fuzzyjoin]{stringdist_left_join}}.
#'
#' @return A data.frame of the same class as that of the one provided in
#'   `trades_directory` and/or `general_directory`. Should `trades_directory` and
#'   `general_directory`be provided as objects of different classes, the class of
#'   the return data.frame will be that of the parent class. i.e. if
#'   `trades_directory` and `general_directory` are provided as a pure data.frame
#'   and a \code{\link[tibble]{tibble}} respectively, a pure data.frame is
#'   returned. Columns include at least `surname`, `forename`,
#'   `address.trade.number`, `address.trade.body`, `address.house.number`,
#'   `address.house.body`.
#'
#' @seealso \code{\link{combine_match_general_to_trades}}.
combine_match_general_to_trades_plain <- function(
  trades_directory, general_directory, verbose, matches, ...
) {

  fun <- function(trades_directory, general_directory, matches, ...) {
    match.string <- NULL

    trades_directory <- combine_make_match_string(trades_directory)
    general_directory <- combine_make_match_string(general_directory) %>%
      dplyr::select(dplyr::matches("^address.house"), match.string)

    combined <- fuzzyjoin::stringdist_left_join(
      x = trades_directory, y = general_directory, by = "match.string", ...
    )

    combined <- if (matches){
      dplyr::select(
        combined, -dplyr::matches("match.string.x"), match = dplyr::matches("match.string.y")
      ) %>%
        dplyr::relocate(match, .after = dplyr::last_col())
    } else { dplyr::select(combined, -dplyr::matches("match.string")) }

    combine_label_failed_matches(combined)

  }

  utils_execute(verbose, fun, trades_directory, general_directory, matches, ...)
}


# combine_match_general_to_trades_progress ####

#' Match general to trades directory records
#'
#' Attempts to complement Scottish post office trades directory data.frame with
#'   house address information from the Scottish post office general directory
#'   data.frame provided by matching records from the two datasets using the
#'   distance metric specified. Shows a progress bar indicating function
#'   progression.
#'
#' @param trades_directory A Scottish post office trades directory in the form
#'   of a data.frame or other object that inherits from the data.frame class
#'   such as a \code{\link[tibble]{tibble}}. Columns must at least include
#'   `surname`, `forename`, `address.trade.number`, `address.trade.body`.
#' @param general_directory A Scottish post office general directory in the form
#'   of a data.frame or other object that inherits from the data.frame class
#'   such as a \code{\link[tibble]{tibble}}. Columns must at least include
#'   `surname`, `forename`, `address.trade.number`, `address.trade.body`,
#'   `address.house.number`, `address.house.body`.
#' @param verbose Whether the function should be executed silently (`FALSE`) or
#'   not (`TRUE`).
#' @param matches Whether (`TRUE`) or not (`FALSE`) a column 'match' showing
#'   general directory matches' name and address(es) should be added to the
#'   output dataset.
#' @param ... Further arguments to be passed down to
#'   \code{\link[fuzzyjoin]{stringdist_left_join}}.
#'
#' @return A data.frame of the same class as that of the one provided in
#'   `trades_directory` and/or `general_directory`. Should `trades_directory` and
#'   `general_directory`be provided as objects of different classes, the class of
#'   the return data.frame will be that of the parent class. i.e. if
#'   `trades_directory` and `general_directory` are provided as a pure data.frame
#'   and a \code{\link[tibble]{tibble}} respectively, a pure data.frame is
#'   returned. Columns include at least `surname`, `forename`,
#'   `address.trade.number`, `address.trade.body`, `address.house.number`,
#'   `address.house.body`.
#'
#' @seealso \code{\link{combine_match_general_to_trades}}.
combine_match_general_to_trades_progress <- function(
  trades_directory, general_directory, verbose, matches, ...
) {

  trades_directory_split <- split(
    trades_directory, (1L:nrow(trades_directory) %/% 100L)
  )

  pb <- progress::progress_bar$new(
    format = "  matching records [:bar] :percent eta: :eta",
    total = length(trades_directory_split), clear = FALSE, width = 100L
  )

  purrr::map_dfr(trades_directory_split, function(df) {
    pb$tick()
    combine_match_general_to_trades_plain(df, general_directory, verbose, matches, ...)
  })
}


# combine_match_general_to_trades ####

#' Match general to trades directory records
#'
#' Attempts to complement Scottish post office trades directory data.frame with
#'   house address information from the Scottish post office general directory
#'   data.frame provided by matching records from the two datasets using the
#'   distance metric specified.
#'
#' @param trades_directory A Scottish post office trades directory in the form
#'   of a data.frame or other object that inherits from the data.frame class
#'   such as a \code{\link[tibble]{tibble}}. Columns must at least include
#'   `surname`, `forename`, `address.trade.number`, `address.trade.body`.
#' @param general_directory A Scottish post office general directory in the form
#'   of a data.frame or other object that inherits from the data.frame class
#'   such as a \code{\link[tibble]{tibble}}. Columns must at least include
#'   `surname`, `forename`, `address.trade.number`, `address.trade.body`,
#'   `address.house.number`, `address.house.body`.
#' @param progress Whether progress should be shown (`TRUE`) or not (`FALSE`).
#' @param verbose Whether the function should be executed silently (`FALSE`) or
#'   not (`TRUE`).
#' @param distance Whether (`TRUE`) or not (`FALSE`) a column 'distance' showing
#'   the string distance between records used for their matching and calculated
#'   using the method specified below should be added to the output dataset.
#' @param matches Whether (`TRUE`) or not (`FALSE`) a column 'match' showing
#'   general directory matches' name and address(es) should be added to the
#'   output dataset.
#' @param ... Further arguments to be passed down to
#'   \code{\link[fuzzyjoin]{stringdist_left_join}}.
#'
#' @return A \code{\link[tibble]{tibble}}; columns include at least `surname`,
#'   `forename`, `address.trade.number`, `address.trade.body`,
#'   `address.house.number`, `address.house.body`.
#'
#' @examples
#' trades_directory <- tibble::tibble(
#'   page = rep("71", 3L),
#'   rank = c("135", "326", "586"),
#'   surname = c("Abbott", "Abercromby", "Blair"),
#'   forename = c("William", "Alexander", "John Hugh"),
#'   occupation = c("Wine and spirit merchant", "Baker", "Victualler"),
#'   type = rep("OWN ACCOUNT", 3L),
#'   address.trade.number = c("18, 20", "12", "280"),
#'   address.trade.body = c("London Road", "Dixon Place", "High Street")
#' )
#' general_directory <- tibble::tibble(
#'   page = rep("71", 2L),
#'   surname = c("Abbott", "Abercromby"), forename = c("William", "Alexander"),
#'   occupation = c("Wine and spirit merchant", "Baker"),
#'   address.trade.number = c("18, 20", ""),
#'   address.house.number = c("136", "29"),
#'   address.trade.body = c("London Road", "Dixon Place"),
#'   address.house.body = c("Queen Square", "Anderston Quay")
#' )
#' combine_match_general_to_trades(
#'  trades_directory, general_directory, progress = TRUE, verbose = FALSE,
#'  distance = TRUE, method = "osa", max_dist = 5
#' )
#'
#' @export
combine_match_general_to_trades <- function(
  trades_directory, general_directory, progress = TRUE, verbose = FALSE,
  distance = TRUE, matches = TRUE, ...
){

  combined <- if (progress)
    combine_match_general_to_trades_progress(
      trades_directory, general_directory, verbose, matches,
      distance_col = "distance", ...
    )
  else
    combine_match_general_to_trades_plain(
      trades_directory, general_directory, verbose, matches,
      distance_col = "distance", ...
    )

  combined <- dplyr::distinct(tibble::tibble(combined))

  if (distance) { combined } else { dplyr::select(combined, -distance) }
}






